h2/src/main/org/h2/mvstore/DataUtils.java

```diff
@@ -681,14 +681,31 @@ private static int parseMapValue(StringBuilder buff, String s, int i, int size)
      * @throws IllegalStateException if parsing failed
      */
     public static HashMap<String, String> parseChecksummedMap(String s) {
-        HashMap<String, String> m = DataUtils.parseMap(s);
-        int check = DataUtils.readHexInt(m, "fletcher", 0);
-        m.remove("fletcher");
-        byte[] bytes = s.getBytes(StandardCharsets.ISO_8859_1);
-        int checksum = DataUtils.getFletcher32(bytes, s.lastIndexOf("fletcher") - 1);
-        if (check == checksum) {
-            return m;
+        HashMap<String, String> map = New.hashMap();
+        StringBuilder buff = new StringBuilder();
+        for (int i = 0, size = s.length(); i < size;) {
+            int startKey = i;
+            i = s.indexOf(':', i);
+            if (i < 0) {
+                throw DataUtils.newIllegalStateException(
+                        DataUtils.ERROR_FILE_CORRUPT, "Not a map: {0}", s);
+            }
+            if (i - startKey == 8 && s.regionMatches(startKey, "fletcher", 0, 8)) {
+                DataUtils.parseMapValue(buff, s, i + 1, size);
+                int check = (int) Long.parseLong(buff.toString(), 16);
+                byte[] bytes = s.getBytes(StandardCharsets.ISO_8859_1);
+                if (check == DataUtils.getFletcher32(bytes, startKey - 1)) {
+                    return map;
+                }
```

Comment:
this assumes that fletcher is always the last key/value in the stored map, do we guarantee that?

Reply:
`parseChecksummedMap()` is used only for chuck footers and storage headers.

Chuck footers contain `fletcher` in the end and padded with spaces and newline. (This padding is removed on reading by old and new code).

Storage headers contain `fletcher` in the end and padded with the single newline character.

So yes, we guarantee this on writing. And it makes sense because both old and new code use only data before `fletcher` record for checksuming.